Index: linux/hpc/migrecv.c
===================================================================
--- linux.orig/hpc/migrecv.c	2006-09-25 20:39:39.000000000 +0200
+++ linux/hpc/migrecv.c	2006-09-25 20:39:41.000000000 +0200
@@ -335,7 +335,7 @@
 
 	return 0;
 
-out:
+/*out:*/
 	printk(KERN_ERR "receive page failed at addr %p\n", (void *) addr);
 	OMBUG("receive page failed at addr %p\n", (void *) addr);
 	__free_page(recv_page);
